### ISSUE TITLE
RFC: Creating CLI API requests from resources

### ISSUE THREAD

**Issue by marpaia** [kolide/fleet#1667]
## Problem

In the [`fleetctl` documentation](https://github.com/kolide/fleet/blob/master/docs/cli/README.md), we outline several file formats that define "resources". Each resource is defined in a file and version controlled (potentially). This issue discusses a proposal of how to translate these files into API requests.

## Considerations

There are two extremes that we've been considering:

### Routing in the CLI

When you execute `fleetctl apply -f ./query.yml`, the `fleetctl` tool will read the file, figure out what type of resource is defined, determine (based on static logic) what endpoint should be requested with what params, etc. 

This is nice because the requests are strongly typed and there is strong client-side validation. Unfortunately, if we add new resources, we have to make sure users have updated both the CLI and the Fleet server.

### Routing on the server

When you execute `fleetctl apply -f ./query.yml`, the `fleetctl` tool will just ship the file content to the server and the server will parse each resource in the file and determine whether or not the content is acceptable.

This is nice because the CLI doesn't need to be updated to take advantage of new resource types, but it's not so great to just have one server endpoint that accepts the text-form of `interface{}`.

## Alternative Proposal

Consider the following resource which defines an osquery query:

```
version: v2
kind: query
spec:
  name: osquery_version
  description: The version of the Launcher and Osquery process
  query: select launcher.version, osquery.version from kolide_launcher_info launcher, osquery_info osquery;
  support:
    launcher: 0.3.0
    osquery: 2.9.0
```

I propose that there is a consistent mapping of the resource metadata to request. For example, the above resource would produce a request such as:

```
POST /api/v2/kolide/query

{
  "name": "osquery_version",
  "description": "The version of the Launcher and Osquery process",
  "query: select launcher.version, osquery.version from kolide_launcher_info launcher, osquery_info osquery;",
  "support": {
    "launcher": "0.3.0",
    "osquery": "2.9.0"
  }
}
```

Basically, the body of the request would be determined by the `spec` data structure. The route would be determined by the `kind` and `version` via `/api/{version}/kolide/{kind}`. The action would determine the HTTP verb (`fleetctl delete` would be `DELETE`, `fleetctl apply` would be `POST`, etc).

This is nice because it allows us to still create a rather nice RESTful API while still allowing for enough dynamicism to allow client version to drift from server version.
